core/src/main/java/io/substrait/relation/ProtoRelConverter.java

```diff
@@ -0,0 +1,286 @@
+package io.substrait.relation;
+
+import io.substrait.expression.AggregateFunctionInvocation;
+import io.substrait.expression.Expression;
+import io.substrait.expression.FunctionLookup;
+import io.substrait.expression.ImmutableExpression;
+import io.substrait.expression.proto.ProtoExpressionConverter;
+import io.substrait.function.SimpleExtension;
+import io.substrait.proto.AggregateFunction;
+import io.substrait.proto.AggregateRel;
+import io.substrait.proto.FetchRel;
+import io.substrait.proto.FilterRel;
+import io.substrait.proto.JoinRel;
+import io.substrait.proto.ProjectRel;
+import io.substrait.proto.ReadRel;
+import io.substrait.proto.SortField;
+import io.substrait.proto.SortRel;
+import io.substrait.type.ImmutableNamedStruct;
+import io.substrait.type.NamedStruct;
+import io.substrait.type.Type;
+import io.substrait.type.proto.FromProto;
+import java.util.ArrayList;
+import java.util.List;
+
+/** Converts from proto to pojo rel representation TODO: AdvancedExtension, CrossJoin, Set */
+public class ProtoRelConverter {
+  static final org.slf4j.Logger logger = org.slf4j.LoggerFactory.getLogger(ProtoRelConverter.class);
+
+  private final FunctionLookup lookup;
+  private final SimpleExtension.ExtensionCollection extensions;
+  private final Type rootType;
```

Comment:
what's rootType?  The root Rel's type? If that's the case, why would we pass rootType to `ProtoExpressionConverter` for the case of `NamedScan`, `VirtualTable`, `emptyScan`?

Reply:
I've removed rootType as a member variable. Instead, I set it based on the context of where I'm calling it from. The two places I don't set it are in the case of a join and in the case of a virtual table. I'd like to pass in whatever ProtoExpressionConverter in all cases, but not sure what it needs.

Reply:
@jacques-n - what should be passed through for `rootType`? It's only used once in ProtoExpressionConverter here:
```
   case ROOT_REFERENCE -> FieldReference.ofRoot(rootType, segments);
```

You end up with an NPE here with a null rootType.

Reply:
If you want to reference a field from an incoming relation in a field reference, the reference not only needs the index, it also needs the data type for that reference. That's the purpose of rootType. It seems unecessary at the global level of rel converter since it would be consumed from one or more inputs of the rel we're currently working on.

Reply:
Do you think it'd make sense to have a two arg constructor for `ProtoExpressionConverter` and we can default rootType to some non null value that doesn't trigger that exception? Or the code could perhaps be tolerant of `rootType` being null?